docs/proposals/role-entry-points.md

```diff
@@ -0,0 +1,264 @@
+# Entry-point based roles
+
+## We have a problem
+
+Each role consists of a set of tasks and variables read/set in some of the tasks.
+Some of the tasks can store their results into registered variables,
+e.g. store stdout of a docker pull command.
+Some of the tasks may be run only under certain condition that are generated
+outside of a role, e.g. pull images only if a deployment is containerized.
+Some of the tasks are run only if a certain variable is defined,
+e.g. add additional Kubelet arguments into the `node-config.yml`
+if `kubelet_additional_args` inventory variable is defined.
+
+Some roles work with a lot of variables that from a point of a role are uninvited citizens.
+Some of the variables are set inside a role, some of them are set outside of a role.
+The variables outside of a role are hard to track, hard to determine their value or
+a condition under which the variables are defined. These unknown facts make a role harder
+to maintain, harder to make a role closed with respect to variables and to make
+a role practically modular.
+
+## Proposal
+
+Make a role that conforms to the following conditions:
+* none of the variables read/set inside a role are global
+* all variables inside a role are `r_` (`r` standing for a role) and/or `l_`(`l` standing for a local) prefixed
```

Comment:
@ingvagabund  I agree that it would be more difficult to tell whether it was passed or created locally but it would be easy for me to determine that it is only used in this role because of the prefix of r_<role name>_<var name>.

I think what @mtnbikenc was saying is that if you set a variable like, `l_is_atomic`  then that is now set globally across all roles.   That means that other roles just inherit that variable with no prior knowledge of it being set.  This encourages the behavior of using it if it is defined.  The issue with this is that the role no longer stands alone.  If we go with a prefix with r_<role_name> then we continue to localize variables to roles and we have a standard variable naming scheme.

I'd probably be ok with another naming scheme if the prefix retained r_<role_name> then added _local_ or something.  If not, I'm not a fan of l_<var name>.